### ISSUE TITLE
multiple routers/services entries no longer working for file provider yaml tcp routes

### ISSUE THREAD

**Issue by schwebke** [traefik/traefik#7715]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I upgraded a working traefik 2.2.11 deployment to 2.3.6.
Part of the configuration are templated tcp routers:
#### `conf/dynamic/tcp-lb1-list.yaml`
```
{{$ports := list "30022" "30122" "30143" "30222" "30322" "30422" "30432" "30522" "30622" "31432" "31532"}}
tcp:
   {{range $i, $p := $ports}}
   routers:
      tcp-lb1-{{$p}}-router:
         entryPoints:
            - "lb1v4t{{$p}}"
            - "lb1v6t{{$p}}"
         rule: "HostSNI(`*`)"
         service: "tcp-lb1-{{$p}}-service"
   services:
      tcp-lb1-{{$p}}-service:
         loadBalancer:
            servers:
               - address: "10.0.0.2:{{$p}}"
               - address: "10.0.0.3:{{$p}}"
               - address: "10.0.0.4:{{$p}}"
   {{end}}
```

### What did you expect to see?
A set of working tcp routes from template expansion, like in 2.2:
![traefik-2 2 11-tcp-routes](https://user-images.githubusercontent.com/6435375/103333301-cd3ab700-4a6d-11eb-9899-afff3b680255.png)

### What did you see instead?
In 2.3 only one route is present:
![traefik-2 3 6-tcp-routes](https://user-images.githubusercontent.com/6435375/103333325-e479a480-4a6d-11eb-81cf-f16c240ffea3.png)

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.3.6
Codename:     picodon
Go version:   go1.15.6
Built:        2020-12-17T16:34:27Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?
- linux amd64
- file provider
- docker

**Comment by schwebke** [comment]
Doing some further investigation without templating, I think the template expansion works. v2.3 does no longer accept multiple routers and services -- so the template loop needs to be pushed one level down.
Not sure if this counts as a bug. It's a breaking change for such configs from 2.2 to 2.3, however.

**Comment by rtribotte** [comment]
Hello @schwebke,

Thanks for your interest in Traefik!

Since the templating is not the cause of your issue, can you rephrase the title?
Also, can you provide a short example of a configuration without templating that works with v2.2 and not with v2.3?

**Comment by schwebke** [comment]
Hi @rtribotte,

here a minimal example resulting in two tcp routes with 2.2, and only one in 2.3:
```
tcp:
   routers:
      tcp-lb1-30022-router:
         entryPoints:
            - "lb1v4t30022"
         rule: "HostSNI(`*`)"
         service: "tcp-lb1-30022-service"
   services:
      tcp-lb1-30022-service:
         loadBalancer:
            servers:
               - address: "10.0.0.2:30022"
               - address: "10.0.0.3:30022"
               - address: "10.0.0.4:30022"
   routers:
      tcp-lb1-30122-router:
         entryPoints:
            - "lb1v4t30122"
         rule: "HostSNI(`*`)"
         service: "tcp-lb1-30122-service"
   services:
      tcp-lb1-30122-service:
         loadBalancer:
            servers:
               - address: "10.0.0.2:30122"
               - address: "10.0.0.3:30122"
               - address: "10.0.0.4:30122"
```

Written like this it does not look good with the duplicate routes/services keys -- using templates it was nice and concise...

I've also attached a minimal docker example to replicate this:
[lb.tar.gz](https://github.com/traefik/traefik/files/5753700/lb.tar.gz)

**Comment by rtribotte** [comment]
Thank you for your quick feedback, it's clearer now.

Since the yaml format must be JSON "valid", duplicates key are illegal.
Yet, you are right that in v2.2 duplicate keys were allowed in the dynamic configuration, thus, this was not intended. It was too permissive.

In the v2.3 version of Traefik, the parsing of the dynamic configuration has evolved and relies now on the paerser library, which not allows duplicate keys. In fact, only the last key applies.

You need to remove the duplicate keys to make your configuration work again.

Unfortunately, I'll close this issue as a question since it's an intended change.
